### ISSUE TITLE
Load pages faster with font-display

### ISSUE THREAD

**Issue by githubjosh** [telusdigital/tds-core#616]
## Description

Currently, our pages cannot render until our TELUS Web fonts have been downloaded. Further, this only happens after parsing the DOM and CSSDOM.

![image](https://user-images.githubusercontent.com/224624/39638340-eca9efc0-4f93-11e8-9c2e-8542850e547f.png)

## Reproduction Percentage

100%. The behaviour above is applicable to every page where we load custom fonts.

## Recommendation

Using the @font-face font-display property, we can control how fonts render in much the same way we can with JavaScript-based solutions. Specifically:

`font-display: swap`

> swap gives the font face a zero second block period and an infinite swap period. This means the browser draws text immediately with a fallback if the font face isn’t loaded, but swaps the font face in as soon as it loads.

Within apps, we will augment this with `<link rel="preload">` to begin downloading each of our .woff2 fonts immediately.

> every browser that supports preloading also supports WOFF2, so that's always the format that you should preload

## References
- https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/webfont-optimization
- https://css-tricks.com/font-display-masses/

## Meta

* Willing to develop solution: Yes
* High impact: Yes

**Comment by ryanoglesby08** [comment]
Thanks for all your efforts Josh! I've reviewed those references and it am distilling what is sounds like the most impactful changes would be in 2 sections, font delivery optimization and page-level optimization.

@githubjosh first question is how to we measure the impact of these changes?

### Page level optimization

> Customize font loading and rendering using `<link rel="preload">`, font-display, or the Font Loading API: default lazyloading behavior may result in delayed text rendering.

Recommend using `font-display` for simplicity, though it is [not supported by IE Edge or IE 11](https://caniuse.com/#feat=css-font-rendering-controls).

@githubjosh why would we need the `<link rel="preload">` strategy if `font-display` is in use? Seems that `font-display: swap` accomplishes the needed behaviour? IE11 does not support either one, though [IE Edge supports preload](https://caniuse.com/#feat=link-rel-preload).

> Give precedence to local() in your src list: listing local('Font Name') first in your src list ensures that HTTP requests aren't made for fonts that are already installed.

We can add `local` to the `@font-face` declarations to skip the HTTP request altogether if the user has `TELUS-Web` already installed. Not sure of the impact here though since we use a custom font that is not expected to already be installed on machines.

### Font delivery optimization

> Specify revalidation and optimal caching policies: fonts are static resources that are infrequently updated. Make sure that your servers provide a long-lived max-age timestamp and a revalidation token to allow for efficient font reuse between different pages. If using a service worker, a cache-first strategy is appropriate.

> Deliver optimized font formats to each browser: provide each font in WOFF2, WOFF, EOT, and TTF formats. Make sure to apply GZIP compression to the EOT and TTF formats, because they are not compressed by default.

For both of these we would need to inspect our CDN for potential optimizations. More analysis needed to determine feasibility, effort, and impact.


### Other

Was talking with @lucylist today about migrating to an SVG based icon set soon too, which would allow us to move off the custom Web font icons we have now.

**Comment by githubjosh** [comment]
@ryanoglesby08 

> how to we measure the impact of these changes?

I would run tds-core in an existing app and include the performance data before and after the implementation. This would include Chrome's Network and Performance data, plus Lighthouse data. Assuming my theory is correct that this will improve page speed, I'd include this analysis in the PR.

> why would we need the <link rel="preload"> strategy if font-display is in use? 

They accomplish different things: Preload starts loading any desired fonts while font-display simply provides an alternative font to render until the desired font has downloaded. Combined, the desired font should be more readily available when it's needed.

**Comment by ryanoglesby08** [comment]
Ok thanks! Makes sense.

Seems that we would want to make sure to include the `type` attribute as well so that browsers will not preload fonts they don't support. (https://www.smashingmagazine.com/2016/02/preload-what-is-it-good-for/)

```
<link rel="preload" href="font1.woff2" as="font" type="font/woff2" crossorigin>
<link rel="preload" href="font1.woff" as="font" type="font/woff" crossorigin>
<link rel="preload" href="font1.ttf" as="font" type="font/ttf" crossorigin>
<link rel="preload" href="font1.eot" as="font" type="font/eot" crossorigin>
```

What is unclear to me is if making a lot of preload requests for un-needed fonts has any downsides? In JS/HTML we don't know which font format the client browser will accept, so we would have to insert a `<link rel="preload">` for all the possible formats for all the fonts (woff2, woff, ttf, and eot for the 4 main fonts = 16 preload requests for fonts).

This seems like it would be less performant/efficient as more network requests are being made overall now, but maybe its fine as preloading does not block page loading. (In the example above, Chrome would only need the woff2 font)

@githubjosh  Is there any downside to making a lot of un-needed preload requests? (There doesn't seem to be from my Googling...)

**Comment by ryanoglesby08** [comment]
Just spoke with @githubjosh , seems like you would only need to preload the woff2 format, because browsers that support preloading also support woff2. 😄 

`@tds/core-css-reset` could expose a list of the fonts, so that starter-kit/all apps can build preload hints without having to hardcode them. (Especially as the font urls will be changing soon)

```js
import { fontUrls } from `@tds/core-css-reset`

// => fontUrls = ['http://cdn.telus.com/font1.woff2', 'http://cdn.telus.com/font1.eot', ...]
```

**Comment by simpleimpulse** [comment]
Fixed in `@tds/core-css-reset@1.1.0`
